Move argument grouping into FunctionSchema

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48687 Delete NativeFunctions.h include from Functions.h
* #48683 Move var and std overloads to Functions.cpp and remove native:: reference
* #48659 Refactor TensorIterator to do allocations via MetaBase::set_output
* **#48195 Move argument grouping into FunctionSchema**
* #48182 Refactor argument fields in FunctionSchema to Arguments
* #48116 Structured kernels generate Meta registrations

The general approach is to change Arguments, splitting `positional`, `kwarg_only` and `out`, into `pre_self_positional`, `self_arg`, `post_self_positional`, and `pre_tensor_options_kwarg_only`, `tensor_options` and `post_tensor_options_kwarg_only`. The splits are as you'd expect: we extract out the self argument and the tensor options arguments, and record the other arguments that came before and after. To do this, we move the logic in `group_arguments` to the parsing process.

Some fuzz in the process:
* I renamed `ThisArgument` to `SelfArgument`, since we don't actually use the terminology "this" outside of C++ (and the model is Python biased)
* I kept the `group_arguments` function, which now just reads out the arguments from the structured model in the correct order. In the long term, we should get rid of this function entirely, but for now I kept it as is to reduce churn.
* I decided to arbitrarily say that when self is missing, everything goes in "post-self", but when tensor options is missing, everything goes in "pre-tensor-options". This was based on where you typically find the argument in question: self is usually at front (so most args are after it), while tensor options are typically at the end (so most args go before it).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25231166](https://our.internmc.facebook.com/intern/diff/D25231166)

ezyang committed:
Move argument grouping into FunctionSchema

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
sorry short on time, no description; coming soon

ezyang committed:
Update on "Move argument grouping into FunctionSchema"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer reviewed (commented):
Just a quick drive-by based on the PR description, front-running a full review later: the choices you're describing in your last bullet point kind of point to the brittleness of modeling the sequence in a maximally partitioned way like this. An alternative (I've pitched this before) is to model self and tensor options as optional pairs of (thing, insertion index) - it's true that the indexes wouldn't be correct by construction, but it avoids the pre-post splintering, and the arbitrary choices (and dynamic checks) for degenerate cases. Still worth modeling args and kwargs as separate lists ofc.

ezyang commented:
@bhosmer Engh, I still don't like it. I gave one reason (tricky bookkeeping if you ever need multiple indices--though tbf we don't have this right now) in our original conversation. But another reason why I don't really like splicing arguments into the lists is that the list prior to splicing isn't really usable for... anything, really.

Imagine:

```
class Arguments:
    kwarg_only: Tuple[Argument, ...]
    tensor_options: Optional[Tuple[TensorOptionsArguments, int]]
```

Does `kwarg_only` contain tensor options arguments or not? You seem to be favoring "make illegal states unrepresentable", so let's say that it doesn't (because if it does, then we need an extra consistency check that the duplicate copy in `kwarg_only` and `tensor_options` agree). So then we have something like this, given the schema `empty.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor`

```
Arguments(
  kwarg_only=("Dimname[]? names", "MemoryFormat? memory_format=None")
  tensor_options=("ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None", 1)
)
```

Use of the `kwarg_only` field directly is useless. There is never a time I want "names" and "memory_format" to be adjacent in an arguments list. I have to do the splice first, or split this into the pre and post parts, before I can do something useful with it. In the pre/post variant, I have pre-done the splitting, because that's what I'm ~always going to want to have happened.

If it's just getting rid of illegal states that we want, the pre/post version can be made to play ball like this:

```
class Arguments:
    pre_kwarg_only: Tuple[Argument, ...]
    tensor_options_and_post_kwarg_only: Optional[Tuple[TensorOptionsArguments, Tuple[Argument, ...]]]
```

but the extra typing here didn't seem worth it to me, at least.

bhosmer commented:
@ezyang thanks for writing out the example, it hits the nail on the head. You're exactly right, the taste/distaste for this approach comes down to whether you feel like the ground representation should be "useful" on its own, or if it's just raw material for derived views. 

Will a full review introduce me to client code that uses the pre/post fragments directly (like in a semantically meaningful way)? If so then I'd agree that what I'm pitching is probably a net loss because it'd have to add derived views for both the full list and the fragments. I'd assumed the full list was the only view that mattered, which I think would put the two representations roughly at parity in terms of view complexity.

You might also be reacting to (per your example) `kwargs_only` being misreadable as a finished list, but I'd think naming and/or commenting would take care of that. But is that part of it?

ezyang commented on tools/codegen/model.py:
```diff
@@ -843,23 +828,95 @@ def __str__(self) -> str:
+        if self.tensor_options is not None:
+            ret.extend(self.tensor_options.all())
+        ret.extend(self.post_tensor_options_kwarg_only)
+        return ret
```
This and positional above are a use of the fields.

ezyang commented on tools/codegen/api/cpp.py:
```diff
@@ -272,46 +272,20 @@ def argument_faithful(
+    args.extend(func.arguments.pre_tensor_options_kwarg_only)
+    if func.arguments.tensor_options is not None:
+        args.append(func.arguments.tensor_options)
+    args.extend(func.arguments.post_tensor_options_kwarg_only)
```
This is another use of the fields, although the intention is to kill this function and make use of the fields directly

ezyang commented:
> Will a full review introduce me to client code that uses the pre/post fragments directly (like in a semantically meaningful way)?

I went ahead and marked the relevant spots. But I didn't really try to refactor anything else after this, so most of the usage is just "squashing" the individual parts back into the list of union which is what the rest of the code uses right now.

Projecting ahead to further refactors, the other things most likely to happen:

1. For methods, you need to drop the self parameter when it doesn't show up in the C++ signature (so actually your rep is good for this)
2. When translating to cpp/dispatcher, depending on `use_c10_full` you will either expand or keep packed the tensor options argument, but it will always be in the same position compared to the other arguments. So splicing will be necessary here
3. Whatever happens to #47712; but most of the heat here is on out arguments so this is essentially the same as (2) (where you are going to do the translation to cpp/dispatcher)

> You might also be reacting to (per your example) kwargs_only being misreadable as a finished list, but I'd think naming and/or commenting would take care of that.

You got me, I was intentionally being a little hyperbolic in the example naming ;)

ezyang committed:
Update on "Move argument grouping into FunctionSchema"


The general approach is to change Arguments, splitting `positional`, `kwarg_only` and `out`, into `pre_self_positional`, `self_arg`, `post_self_positional`, and `pre_tensor_options_kwarg_only`, `tensor_options` and `post_tensor_options_kwarg_only`. The splits are as you'd expect: we extract out the self argument and the tensor options arguments, and record the other arguments that came before and after. To do this, we move the logic in `group_arguments` to the parsing process.

Some fuzz in the process:
* I renamed `ThisArgument` to `SelfArgument`, since we don't actually use the terminology "this" outside of C++ (and the model is Python biased)
* I kept the `group_arguments` function, which now just reads out the arguments from the structured model in the correct order. In the long term, we should get rid of this function entirely, but for now I kept it as is to reduce churn.
* I decided to arbitrarily say that when self is missing, everything goes in "post-self", but when tensor options is missing, everything goes in "pre-tensor-options". This was based on where you typically find the argument in question: self is usually at front (so most args are after it), while tensor options are typically at the end (so most args go before it).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer commented on tools/codegen/model.py:
```diff
@@ -902,12 +960,85 @@ def parse(args: str) -> 'Arguments':
+            # If there is enough space...
+            if i <= len(kwarg_only) - len(predicates):
+                # And the next len(predicates) arguments look like TensorOptions arguments
+                if all(p(a) for p, a in zip(predicates, kwarg_only[i : i + len(predicates)])):
```
It still feels like a small devex pothole to be failing silently on near misses here, but probably out of scope for this work.

ezyang replied:
Yeah, this is preexisting code, so out of scope for this PR. Note that we do have plenty of "near-ish misses" (e.g., one or two arguments from TensorOptions) in various functions, including 

```
- func: to.device(Tensor self, Device device, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor
  use_c10_dispatcher: full
  variants: method
  device_guard: False
```

so it's not altogether clear to me what a "you mistyped this" error message would look like.

bhosmer replied:
Yeah, the line isn't trivially easy to draw. I guess if the goal state is "nobody uses TensorOptions for factory functions in the C++ API", then the most robust setup might be to drive the gather logic from an explicit annotation in native_functions rather than a pattern match - could even just use `category_override: factory`.

bhosmer reviewed (approved):
LGTM!

ezyang committed:
Update on "Move argument grouping into FunctionSchema"


The general approach is to change Arguments, splitting `positional`, `kwarg_only` and `out`, into `pre_self_positional`, `self_arg`, `post_self_positional`, and `pre_tensor_options_kwarg_only`, `tensor_options` and `post_tensor_options_kwarg_only`. The splits are as you'd expect: we extract out the self argument and the tensor options arguments, and record the other arguments that came before and after. To do this, we move the logic in `group_arguments` to the parsing process.

Some fuzz in the process:
* I renamed `ThisArgument` to `SelfArgument`, since we don't actually use the terminology "this" outside of C++ (and the model is Python biased)
* I kept the `group_arguments` function, which now just reads out the arguments from the structured model in the correct order. In the long term, we should get rid of this function entirely, but for now I kept it as is to reduce churn.
* I decided to arbitrarily say that when self is missing, everything goes in "post-self", but when tensor options is missing, everything goes in "pre-tensor-options". This was based on where you typically find the argument in question: self is usually at front (so most args are after it), while tensor options are typically at the end (so most args go before it).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move argument grouping into FunctionSchema"


The general approach is to change Arguments, splitting `positional`, `kwarg_only` and `out`, into `pre_self_positional`, `self_arg`, `post_self_positional`, and `pre_tensor_options_kwarg_only`, `tensor_options` and `post_tensor_options_kwarg_only`. The splits are as you'd expect: we extract out the self argument and the tensor options arguments, and record the other arguments that came before and after. To do this, we move the logic in `group_arguments` to the parsing process.

Some fuzz in the process:
* I renamed `ThisArgument` to `SelfArgument`, since we don't actually use the terminology "this" outside of C++ (and the model is Python biased)
* I kept the `group_arguments` function, which now just reads out the arguments from the structured model in the correct order. In the long term, we should get rid of this function entirely, but for now I kept it as is to reduce churn.
* I decided to arbitrarily say that when self is missing, everything goes in "post-self", but when tensor options is missing, everything goes in "pre-tensor-options". This was based on where you typically find the argument in question: self is usually at front (so most args are after it), while tensor options are typically at the end (so most args go before it).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25231166](https://our.internmc.facebook.com/intern/diff/D25231166)

[ghstack-poisoned]
